### PR TITLE
Switch to all in one workflow strategy for GitHub Actions

### PR DESCRIPTION
**Pull request by abelikt**
## Proposed changes

Switch to all in one workflow strategy for GitHub Actions.
See  https://github.com/thin-edge/thin-edge.io/issues/1038

Note: Docu is still missing (Runners, configuration, etc)
Note: The configuration of Runners is not finalized yet

**Note: I'll present this in the next arch rampup**

Here is an example how it would look on Michaels CI:

https://github.com/abelikt/thin-edge.io/actions/runs/2108411912

This is the report that is automatically generated from it:

https://abelikt.github.io/thin-edge.io/report-matrix.html

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/1038

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

comment

### CODE REVIEW

File: .github/workflows/ci_pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,1366 @@
...
+        continue-on-error: true
+
+      - name: Stop apama
+        run: sudo systemctl stop apama
+        continue-on-error: true
+
+      - name: purge
+        run: sudo dpkg -P tedge_agent tedge_logfile_request_plugin tedge_mapper tedge_apt_plugin tedge_apama_plugin tedge mosquitto-clients mosquitto libmosquitto1 collectd-core collectd
+
+      - name: install mosquitto
```
[the remark this answers is not in the document]
Comment (abelikt) [reply .github/workflows/ci_pipeline.yml:813]: I absolutely agree and I would propose to do this in this story here : https://github.com/thin-edge/thin-edge.io/issues/644

File: .github/workflows/ci_pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,1366 @@
...
+        # https://github.com/marketplace/actions/rust-cargo
+        with:
+          command: test
+          args: --no-fail-fast
+
+
+###############################################################################
+
+  build_amd64:
+    name: Build tedge and mapper Debian packages for amd64
```
[the remark this answers is not in the document]
Comment (abelikt) [reply .github/workflows/ci_pipeline.yml:135]: I absolutely agree and I would propose to do this in this story here : https://github.com/thin-edge/thin-edge.io/issues/644

File: .github/workflows/ci_pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,1384 @@
...
+# TODO: Think about automated generation of this workflow, there is
+#   duplicated code, especially in the later sections
+# TODO: Smoke testing on Azure is disabled, we should check if it works out
+#   with 4 runners in parallel or enable only one
+
+on:
+  push:
+    branches: [main]
+  workflow_dispatch:
+    branches: [main]
```
Comment (rina23q) [inline .github/workflows/ci_pipeline.yml:18]: I am wondering if the specifying `main` in `workflow_dispatch` is meaningful. 

P.S. I just realized that `workflow_dispatch` can accept user input and can use them as environmental values. https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch

File: .github/workflows/ci_pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,1384 @@
...
+    runs-on: Ubuntu-20.04
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Run black
+        uses: rickstaa/action-black@v1
+        with:
+          black_args: ". --check"
```
Comment (rina23q) [inline .github/workflows/ci_pipeline.yml:57]: It seems good to have this step on every PR.

File: .github/workflows/ci_pipeline.yml [hunk 5]
```diff
@@ -0,0 +1,1384 @@
...
+        uses: actions-rs/cargo@v1
+        # https://github.com/marketplace/actions/rust-cargo
+        with:
+          command: test
+          args: --verbose --features integration-test,requires-sudo -- \
+            --skip sending_and_receiving_a_message
+
+#################################################################################
+
+  install-and-use-rpi_m32sd10a:
```
Comment (rina23q) [inline .github/workflows/ci_pipeline.yml:535]: The code duplication looks nightmare to maintain in the future :/

**Review by rina23q — APPROVED** [review]

**Comment by rina23q** [comment]
I prefer to merge this PR first.
1) Keep the existing workflows and this all-in-one workflow.
2) Remove the duplication and complexity of the all-in-one workflow. #644.
3) Once the all-in-workflow gets stable, we will remove the existing workflows and keep only the all-in-one workflow.

**Pull request merged**